gix-object/src/lib.rs

```diff
@@ -261,13 +261,13 @@ pub mod decode {
         use crate::bstr::{BString, ByteSlice};
 
         /// The type to be used for parse errors.
-        pub type ParseError<'a> = nom::error::VerboseError<&'a [u8]>;
+        pub type ParseError<'a> = winnow::error::VerboseError<&'a [u8]>;
```

Comment:
What is the goal with `verbose-object-parsing-errors`?

`VerboseError` is gone in 0.5, replaced with either
- `ContextError<Context>` where the default context allows `Label` (description of what you are parsing) and `Expected` (what is needed to parse it)
- `TreeError` which tracks all error state and is more useful for tests as the extra information might help when debugging

We could
- Use `ContextError<&'static str>` and provide our own rendering
- Use `ContextError<StrContext>` and switch the code to use `StrContext` instead of `&'static str`
- Create a custom error type

Reply:
Thanks for bringing this up.

The original thinking here was that object parsing should be as fast as possible, so reducing errors to booleans seemed like the right path to take. I probably tested this and measured 1% performance improvement that way, but my memory is fuzzy.

With that said, I wouldn't mind going with 'high-performance' errors by default that leave the possibility to replace the error later with `()` or something similarly light. That way, one could benchmark it once more and see if it's worth the complexity to maintain both versions of the error.

Ideally, there would be no split and errors would be enabled and helpful if parsing fails, but fast enough to not make us want to disable them. I have a feeling that parsing is a little faster anyways now which leaves room for having decent errors without being slower than the no-error `nom` version in the end.

I'd absolutely leave it to your expertise to chose the right error type, and I will be sure to benchmark it against the no-error version on `main`. From there we can see if it can just stay like this.

For reference, `Commit` parsing has to be very fast, and tree-parsing already has its own fast-path anyway that implements it by hand. Commits are the only object type I am thinking about.

Reply:
Good to know; thanks.

- `basic` is 0.4's `Error` which was renamed to `InputError`
- `unit` is `()` (and hadn't previously been benchmarking it)
- `context` is 0.5's `ContextError` which replaced `VerboseError` and was pulled from `toml_edit`
```console
$ cargo bench --bench json -- canada # v0.4.9
json/basic/canada       time:   [10.473 ms 10.498 ms 10.525 ms]
json/verbose/canada     time:   [31.411 ms 31.488 ms 31.574 ms]

$ cargo bench --bench json -- canada # v0.5.7
json/basic/canada       time:   [9.9094 ms 9.9743 ms 10.040 ms]
json/unit/canada        time:   [8.7462 ms 8.7724 ms 8.7998 ms]
json/context/canada     time:   [14.040 ms 14.082 ms 14.128 ms]
```

In short, `()` still offers a significant performance improvement though `ContextError` does close the gap quite a bit compared to `VerboseError`.

Reply:
The purpose of the error by default would just be to report error with boolean precision, but ideally there is a way to swap-in/toggle-on the best-possible error to help understanding what's wrong with a particular kind of object.

Thinking about it, it would be even better if this functionality could exist in the same binary (now that we are here to improve things). The workflow could be that parsing is very fast by default, anticipating no issues, but if there is an issue, the object could be parsed again with the sole purpose of generating an error to pinpoint the source of the issue.

Ultimately, I trust your judgement on this even if it means choosing one kind of error that isn't unit, because it might still be faster than is on `main` and I will be sure to test that.